test/framework/exec/docker.go

```diff
@@ -0,0 +1,398 @@
+	}
+
+	execConfig := types.ExecConfig{
+		Privileged:   true,
+		Cmd:          append([]string{command}, args...),
+		AttachStdout: true,
+		AttachStderr: true,
+	}
+
+	response, err := docker.ContainerExecCreate(ctx, containerName, execConfig)
+	if err != nil {
+		return err
+	}
+
+	execID := response.ID
+	if execID == "" {
+		return fmt.Errorf("exec ID empty")
+	}
+
+	resp, err := docker.ContainerExecAttach(ctx, execID, types.ExecStartCheck{})
+	if err != nil {
+		return err
+	}
+	defer resp.Close()
+
+	// Read out any output from the call
+	outputErrors := make(chan error)
+	go func() {
+		// Send the output to the host file
+		_, err = stdcopy.StdCopy(fileOnHost, fileOnHost, resp.Reader)
+		outputErrors <- err
+	}()
+
+	select {
+	case err := <-outputErrors:
+		if err != nil {
+			return err
+		}
+
+	case <-ctx.Done():
+		return err
+	}
+
+	return nil
+}
+
+// ContainerRunConfig holds the configuration settings for running a container.
+type ContainerRunConfig struct {
+	// Image is the name of the image to run.
+	Image string
+	// Network is the name of the network to connect to.
+	Network string
+	// User is the user name to run as.
+	User string
+	// Group is the user group to run as.
+	Group string
+	// Volume is a collection of any volumes (docker's "-v" arg) to mount in the container.
+	Volumes map[string]string
+	// EnvironmentVars is a collection of name/values to pass as environment variables in the container.
+	EnvironmentVars map[string]string
+	// CommandArgs is the command and any additional arguments to execute in the container.
+	CommandArgs []string
+}
+
+// OwnerAndGroup gets the user configuration for the container (user:group).
+func (crc *ContainerRunConfig) OwnerAndGroup() string {
+	if crc.User != "" {
+		if crc.Group != "" {
+			return fmt.Sprintf("%s:%s", crc.User, crc.Group)
+		}
+
+		return crc.User
+	}
+
+	return ""
+}
+
+// EnvironmentVariables gets the collection of environment variables for the container.
+func (crc *ContainerRunConfig) EnvironmentVariables() []string {
+	envVars := []string{}
+	for key, val := range crc.EnvironmentVars {
+		envVars = append(envVars, fmt.Sprintf("%s=%s", key, val))
+	}
+	return envVars
+}
+
+// Bindings gets the volume mount bindings for the container.
+func (crc *ContainerRunConfig) Bindings() []string {
+	bindings := []string{}
+	for src, dest := range crc.Volumes {
+		bindings = append(bindings, volumeMount(src, dest))
+	}
+	return bindings
+}
+
+// RunContainer will run a docker container with the given settings and arguments, returning any errors.
+func RunContainer(ctx context.Context, runConfig *ContainerRunConfig, output io.Writer) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	containerConfig := container.Config{
+		Tty:          true, // allocate a tty for entrypoint logs
+		Image:        runConfig.Image,
+		Cmd:          runConfig.CommandArgs,
+		AttachStdout: true,
+		AttachStderr: true,
+	}
+	user := runConfig.OwnerAndGroup()
```

Comment:
I think we can also just set it in the struct literal in l.237-l.242. Because if the user string is empty it won't change anything in the field, so it's not necessary to handle the empty case.